src/nl/hannahsten/texifyidea/inspections/latex/probablebugs/LatexQuoteInspection.kt

```diff
@@ -0,0 +1,118 @@
+     */
+    data class QuoteMaker(val command: String, val openParam: Int, val closeParam: Int)
+
+    /**
+     * All the ways that csquotes defines a new quotation mark
+     */
+    private val quoteMakers = arrayOf(
+        QuoteMaker("\\MakeOuterQuote", 0, 0),
+        QuoteMaker("\\MakeInnerQuote", 0, 0),
+        QuoteMaker("\\MakeAutoQuote", 0, 1),
+        QuoteMaker("\\MakeForeignQuote", 1, 2),
+        QuoteMaker("\\MakeBlockQuote", 0, 2),
+        QuoteMaker("\\MakeForeignBlockQuote", 1, 3),
+        QuoteMaker("\\MakeHyphenBlockQuote", 1, 3),
+        QuoteMaker("\\MakeHybridBlockQuote", 1, 3),
+    );
+
+    override fun inspectFile(file: PsiFile, manager: InspectionManager, isOntheFly: Boolean): List<ProblemDescriptor> {
+        // Build up a list of acceptable quote pairs. The ith index in both arrays must contain a matching quote
+        // pair. Additional quote pairs are extracted from the csquotes quote making commands
+        val openers = LinkedHashSet(arrayListOf("`", "``"));
+        val closers = LinkedHashSet(arrayListOf("'", "''"));
+        for (quoteMaker in quoteMakers) {
+            for (command in file.commandsInFile(quoteMaker.command)) {
+                val opener = command.requiredParameter(quoteMaker.openParam);
+                val closer = command.requiredParameter(quoteMaker.closeParam);
+
+                if (!opener.isNullOrEmpty() && !closer.isNullOrEmpty()) {
+                    openers.add(opener);
+                    closers.add(opener);
+                }
+            }
+        }
+
+        // Process each of the quote tokens individually
+        val quoteStack = ArrayDeque<Int>();
+        val issues = descriptorList();
+        for (text in findChildrenOfType(file, LatexNormalText::class.java)) {
+            for (match in quotePattern.findAll(text.text)) {
+                val openIndex = openers.indexOf(match.value);
+                if (openIndex > -1) {
+                    // First case: we find a valid opener
+                    // In that case, push the ID of that opener onto the stack
+                    quoteStack.add(openIndex);
+                }
+                else {
+                    val closeIndex = closers.indexOf(match.value);
+                    if (closeIndex == -1) {
+                        // Second case: we find a token that's neither a valid opener nor closer
+                        issues.add(
+                            manager.createProblemDescriptor(
+                                text,
+                                match.range.toTextRange(),
+                                "${match.value} is not a valid set of LaTex quotes",
+                                ProblemHighlightType.GENERIC_ERROR_OR_WARNING,
+                                isOntheFly
+                            )
+                        )
+                    }
+                    else if (!quoteStack.isEmpty() && closeIndex == quoteStack.last()) {
```

Comment:
So `closeIndex == quoteStack.last()` is supposed to denote that these opening and closing quotes belong to the same pair, right? Maybe it would be more readable if it read `quotesPairIndex == quotesPairIndexStack.last()` or something that indicates that these are indices in the list of quote pairs.

Reply:
Yes that's right, they are indeed indices, so I'll change this. I'm just used to Python where we try to be concise rather than precise.

Reply:
Actually I always tell Python developers to be more verbose in variable names, because you have autocompletion (and GitHub Copilot now as well) and Python is extremely verbose anyway so go with the Pythonic flow of verbosity.